### ISSUE TITLE
Memory leak with scaleoffset

### ISSUE THREAD

**Issue by a6** [h5py/h5py#984]
Hello,

First, thank you for a phenomenal library. It is amazing what h5py does for large datasets.

I believe I have identified a memory leak when using `scaleoffset` on chunked datasets. The leak exists with datasets of dtypes integers and floats. For both dataset dtypes, the leak is eliminated when `scaleoffset` is not used.

For integers:

Leak is present when `scaleoffset=True/0`. The leak is not present for positive values of `scaleoffset`. A simple workaround here is to never use `scaleoffset=0/True`, but the spirit of lossless use of `scaleoffset` without needing to count bits is very attractive.

Minimum test case:
```python
import numpy, h5py

dsize = 10**12
lim = 2**7 - 1
bl = 1000 # block size

f = h5py.File('./test.h5', 'w')
d = f.create_dataset('d', shape=(dsize, ), dtype=numpy.int8, chunks=True, scaleoffset=True)

for i in range(int(dsize/bl)) :
    d[i*bl : (i+1)*bl] = numpy.random.randint(low=-lim, high=lim, size=bl, dtype=numpy.int8)
```
On the system I tested with, this grows the process VM size by around 10MB/sec until memory is exhausted and the process is killed by Linux. Speed of memory consumption appears to scale with the number of bytes used in the dtype (`int16` consumes 30MB/s, `int32` consumes 60MB/s, `int64` consumes ~115MB/s). Leak is present for unsigned integers as well.

For floats:

The leak with floats is more subtle, as certain values for `scaleoffset` work, but others do not. I have not exhaustively tested this, but there seems to be some kind of precision threshold beyond which the leak emerges.

Minimum test case:
```python
import numpy, h5py

dsize = 10**12
lim = 10**5
bl = 1000 # block size

f = h5py.File('./test.h5', 'w')
d = f.create_dataset('d', shape=(dsize, ), dtype=numpy.float32, chunks=True, scaleoffset=5)

for i in range(int(dsize/bl)) :
    d[i*bl : (i+1)*bl] = numpy.random.uniform(low=-lim, high=lim, size=bl)
```
In this example, `scaleoffset=5` is the cutoff for the `float32` dataset to cause a leak of ~35MB/s. If `scaleoffset=4`, there is no leak. If the dtype is switched to `float64` in this example, the threshold to leak appears at `scaleoffset=14` with a leak of ~72MB/s, below which memory is stable.

Test performed on:
 * x86_64 Ubuntu Xenial (16.04)
 * Python 3.5.2 (from Ubuntu repository)
 * numpy and h5py pulled from latest PyPi repository
 * h5py.version.info:
```
h5py    2.7.1
HDF5    1.8.18
Python  3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.14.0
```
 * uname -a:
`Linux dev 4.4.0-109-generic #132-Ubuntu SMP Tue Jan 9 19:52:39 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

Thank you for any time the team may devote to solving this.

**Comment by a6** [comment]
I should have also mentioned that the issue is present with numpy 1.13.3 as well.

**Comment by aragilar** [comment]
@a6 Would you be able to test this on HDF5 1.10? Also, have you seen this with previous h5py versions (do you know if this is a regression, or has this leak always been there)?

**Comment by a6** [comment]
Is there a launchpad/ubuntu repository with HDF5 1.10 for xenial? If so, it's very easy for me to test with a different HDF5 library. If not, it's quite a bit harder to figure out as I'd have to figure out how to build h5py against a different library path, presumably?

**Comment by a6** [comment]
On the other question, this is the only version of h5py I've been playing with. I don't know if this is new behavior.
There are similar reports of strange behavior with chunked datasets from years back, but I am not sure if it's the same problem:
https://groups.google.com/forum/#!topic/h5py/_a35vzQzRrg

**Comment by chrisjbillington** [comment]
I have the same problem as the chunked datasets one, with the version of hdf5 1.10 provided by the Ubuntu 18.04 repositories. This code:

```
import h5py

def foo():
    with h5py.File('schro.h5','r') as f:
        psi = f['psi'][:]

foo()
```

loads a 11GB dataset, but once the relevant variables go out of scope, no matter what I do, there's 7GB of additional memory I can't get back, even if  I call `gc.collect()`.

As far as I can tell, if I open the *same* dataset again, it doesn't leak an additional 7GB. But if I open a *different* dataset, it does.

I don't know how big my dataset will be in advance, so I can't make a fixed dataset size at creation time, instead I just keep appending to a dataset. But I run out of my 32GB of memory during the analysis script that runs afterward. My solution for the moment is to do h5repack with the `-l CONTI` option before the analysis, this seems to resolve the memory leak.

**Comment by tacaswell** [comment]
This should be reported to upstream to hdf5.

h5py just passes `scaleoffset` through to the filter chain the hdf5 does on the way in and out. That repack helps with this is pretty strong corroboration.

**Comment by aparamon** [comment]
Upstream report:
https://jira.hdfgroup.org/browse/HDFFV-10705

**Comment by takluyver** [comment]
I'm going to drop the milestone because this is being fixed upstream.

@aparamon do you think h5py needs to do anything to take advantage of the fix, or can we just close it?

**Comment by a6** [comment]
Is there anyone who can view the status of that upstream report and provide an update? I don't have access to the HDF5 JIRA. Thank you!

**Comment by epourmal** [comment]
It was resolved in 1_10 and develop branches https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse and will be available in HDF5 1.12.0 (end of July 2019) and in 1.10.6 (November 2019).

To get access to HDF5 Jira, please register with the The HDF Group website (https://hdfgroup.org)

Thank you!
Elena

**Comment by takluyver** [comment]
I'll close this as it doesn't sound like any change is needed in h5py to benefit from the fix upstream. We can reopen it if someone knows that a change is needed.
